src/common/base/test/RangeVsTransformBenchmark.cpp

```diff
@@ -0,0 +1,130 @@
+    std::vector<int32_t> from;
+    std::vector<int32_t> to;
+    BENCHMARK_SUSPEND {
+        from.resize(1000, 0);
+        to.resize(1000);
+    }
+    int32_t index = 0;
+    for (auto& i : from) {
+        to[index++] = i;
+    }
+    folly::doNotOptimizeAway(to);
+}
+BENCHMARK_RELATIVE(Test2_TransformInt) {
+    std::vector<int32_t> from;
+    std::vector<int32_t> to;
+    BENCHMARK_SUSPEND {
+        from.resize(1000, 0);
+        to.resize(1000);
+    }
+    std::transform(from.begin(), from.end(), to.begin(), [] (const auto& e) {
+        return e;
+    });
+    folly::doNotOptimizeAway(to);
+}
+
+BENCHMARK_DRAW_LINE();
+
+BENCHMARK(Test3_RangeTestInt) {
+    std::vector<int32_t> v;
+    BENCHMARK_SUSPEND {
+        v.resize(1000, 0);
+    }
+    for (auto& i : v) {
+        (void)(i);
+        int a;
+        folly::doNotOptimizeAway(a);
+    }
+}
+
+BENCHMARK_RELATIVE(Test3_ForEachInt) {
+    std::vector<int32_t> v;
+    BENCHMARK_SUSPEND {
+        v.resize(1000, 0);
+    }
+    std::for_each(v.begin(), v.end(), [] (const auto&) {
+        int a;
+        folly::doNotOptimizeAway(a);
+    });
+}
+
+int main(int argc, char** argv) {
+    folly::init(&argc, &argv, true);
+    folly::runBenchmarks();
+    return 0;
+}
+
+
+/*
+-O2
+[heng.chen@hp-server nebula]$ src/common/base/test/_build/range_vs_transform__bm
+============================================================================
+/data/heng/workspace/nebula/src/common/base/test/RangeVsTransformBenchmark.cpprelative  time/iter  iters/s
+============================================================================
+Test1_RangeTestStr                                          80.60us   12.41K
+Test1_TransformStr                               100.47%    80.22us   12.47K
+----------------------------------------------------------------------------
+Test2_RangeTestInt                                         713.67ns    1.40M
+Test2_TransformInt                                99.99%   713.72ns    1.40M
+----------------------------------------------------------------------------
+Test3_RangeTestInt                                         383.15ns    2.61M
+Test3_ForEachInt                                 100.63%   380.75ns    2.63M
+============================================================================
+
+-O3
+[heng.chen@hp-server nebula]$ src/common/base/test/_build/range_vs_transform__bm
+============================================================================
+/data/heng/workspace/nebula/src/common/base/test/RangeVsTransformBenchmark.cpprelative  time/iter  iters/s
+============================================================================
+Test1_RangeTestStr                                          84.19us   11.88K
+Test1_TransformStr                               100.01%    84.18us   11.88K
+----------------------------------------------------------------------------
+Test2_RangeTestInt                                         222.70ns    4.49M
```

Comment:
Please expain why `-O3` is better than `-O2` for this case.

Reply:
Let me try it.

Reply:
Generate the asm for the snippet
```
int from[100000];
int to[100000];

int f() {
    int index = 0;
    for (auto& i : from) {
        to[index++] = i;
    }
    return 0;
}
```
With -O2
```
.LFB1898:
	.cfi_startproc
	xorl	%eax, %eax
	.p2align 4,,10
	.p2align 3
.L2:
	movl	from(%rax), %edx
	addq	$4, %rax
	movl	%edx, to-4(%rax)
	cmpq	$400000, %rax
	jne	.L2
	xorl	%eax, %eax
	ret
	.cfi_endproc
```
With -O3
```
_Z1fv:
.LFB1898:
	.cfi_startproc
	subq	$8, %rsp
	.cfi_def_cfa_offset 16
	movl	$400000, %edx
	movl	$from, %esi
	movl	$to, %edi
	call	memcpy
	xorl	%eax, %eax
	addq	$8, %rsp
	.cfi_def_cfa_offset 8
	ret
	.cfi_endproc
```

Notice that,  with O3,  there is no loop,  just call memcpy directly,  very smart!

Reply:
Excellent.